rfcs/008-smart-transfer-protocol.md

```diff
@@ -0,0 +1,270 @@
+   - the negotiation can repeat for another subset of missing object files
+- Client commits the new synchronized blocks, and closes the operation
+
+For a `push` operation:
+- Client initiates the request to the Server, sending the ID of the dataset and a tarball of the new metadata blocks
+- If Server does not detect a divergence in the metadata, it confirms the Client may proceed synchronizing objects
+- Client negotiates with the Server on the upload method for a subset or full set of new object files:
+   - for each object file, Server creates a time-limited pre-signed upload URL with write-only access to the storage system
+   - Client directly uploads the object files to the storage system via pre-signed URLs
+   - the negotiation can repeat for another subset of new object files
+- Client informs the Server about completion of the objects synchronization
+- Server verifies the required objects exist and attempts to commit the new metadata blocks
+- In case of success, the Server confirms the commit succeeded, and closes the operation
+
+The exchange of object files should be performed with delays and object sizes taken into account.
+As the expiration time of pre-signed URLs is limited, in case of large bulks have to be exchanged,
+the pre-signed URLs might expire before the actual transfer starts - in this case they need to be re-requested.
+
+The parallelism of the object download/upload operations to/from storage system is not limited by the protocol,
+neither is to be constrained by the server. It's considered to be an implementation detail of the Client.
+
+In case of errors, timeouts, the Client needs to re-initiate the operation with the Server by opening another session.
+New operation may skip the synchronization of the already processed object files from the previous sessions, 
+even if the changes have not been finally committed to the metadata. Cleaning orphan files that accumulate on transfer errors
+is a subject for a separate Garbage Collector mechanism within the dataset storage, that is out of the scope of this RFC.
+
+The protocol can be implemented using [WebSockets](https://websockets.spec.whatwg.org//) or any other 
+alternative type of channel that supports bi-directional non-blocking message exchanges.
+
+## Reference-level explanation
+
+See `AsyncAPI` [description of the protocol messages](../swagger/smart_transfer_protocol_ws.yaml).
+
+### Push flow
+
+![Diagram: ODF Smart TransferProtocol Push](../images/smart_transfer_protocol_push.svg)
+
+### Pull flow
+
+![Diagram: ODF Smart TransferProtocol Pull](../images/smart_transfer_protocol_pull.svg)
+
+
+## Drawbacks
+
+
+## Rationale and alternatives
+
+
+## Prior art
+
+- [Git's "Smart Protocol"](https://git-scm.com/book/en/v2/Git-Internals-Transfer-Protocols)
+
+## Unresolved questions
+- Garbage Collector mechanism within the dataset storage
+
+
+## Future possibilities
+
+
+## Unstructured original notes
```

Comment:
I think we should distill this into other sections rather than leaving this as a pile of notes, e.g.

**Drawbacks:**
- Ended up with a custom protocol rather than reusing some standard

**Prior art:**
- git smart transfer protocol
- DeltaSharing + good/bad notes 

**Unresolved questions:**
- Should this protocol support live streaming of datasets?
- Should it support in-band data transfers?
  - And if yes, should it be used for query results or for getting data from projections

**Future possibilities:**
- DeltaSharing can be implemented as a separate protocol reusing a lot of the parts from this
- A detection of unfinished pushes can be added to clean up orphaned files without doing a full GC run on a dataset